### ISSUE TITLE
Batch Crafting Causes Crash Upon Using All Available Ingredients in Batch

### ISSUE THREAD

**Issue by Taichi22** [CleverRaven/Cataclysm-DDA#37239]
[Debug Zip.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4084629/Debug.Zip.zip)

Save file, config file, and debug log all zipped into file provided.

# Describe the bug

Attempts to make troglobite serum result in an error wherein the recipe ingredients are consumed to create an object which can be activated but upon doing so say that the required components to continue crafting are unavailable.

# Steps To Reproduce

Unsure of how to do so in other save games, however, in uploaded save:
1. Begin crafting a single unit of troglobite serum.
2. Halt craft using period key.
3. Batch-craft all other available units of troglobite serum as a single craft.

# Expected behavior

Logically, the batch, upon being created, should allow crafting, at least until one or more ingredients have been wasted during the crafting process and further ingredients are required to replace the wasted ones. 
Alternatively, it would make sense to warn the player about the fact that failed batches will be halted until more ingredients are gotten, and/or break down any in-progress items and eject all results from the current crafting state should batch-crafting run out.

# Screenshots

![Capture](https://user-images.githubusercontent.com/1211989/72702152-aff02f80-3b1f-11ea-9316-b489df2b712b.JPG)
Error message thrown after crafting, pausing, then batch-crafting.

![Capture1](https://user-images.githubusercontent.com/1211989/72702213-d910c000-3b1f-11ea-86da-ebe13078a6de.JPG)
Issue caused when _only_ batch crafting and using all ingredients.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.D-11480-gdcdef1e [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Folding Parts pack [deoxymod],
    Vehicle Additions Pack [blazemod],
    Tanks and Other Vehicles [Tanks],
    Mythical Martial Arts [MMA],
    Hydroponics [hydroponics],
    EZ-Mode Medical [fast_healing],
    Modular Turrets [modular_turrets],
    Salvaged Robots [Salvaged_Robots],
    Mythological Replicas [nw_pack],
    More Survival Tools [More_Survival_Tools],
    Medieval and Historic Classes and Shields [Medieval_Stuff],
    Makeshift Items Mod [makeshift],
    Icecoon's Arsenal [ew_pack],
    Garden Pots [growable-pots],
    DeadLeaves' Fictional Guns [FIC_Weapons],
    Craftable Gun Pack [craftgp],
    C.R.I.T Expansion Mod [crt_expansion],
    Aftershock [aftershock],
    Graphical Overmap [Graphical_Overmap],
    Stats Through Skills [StatsThroughSkills],
    Stats Through Kills [stats_through_kills],
    Safe Autodoc [safeautodoc],
    Manual Bionic Installation [manualbionicinstall]
]

 - OS: [e.g. iOS 8 or Windows 10 or Ubuntu 18.04]
 - Game Version: [from the main menu, e.g. 0.C-29938-g90f5268437]
 - Graphics version: [Tiles or Terminal]
 - Mods loaded: [e.g.dda, boats, hacktheplanet, StatsThroughSkills]

# Additional context

Blacklist removed from Modular Turrets [modular_turrets], and Salvaged Robots [Salvaged_Robots] due to balance issues with control laptop in current experimental. (I realize that control laptop is unbalanced, but not being able to make one is also horrendously debilitating when EMP bombs are nigh-impossible to make as well.)

Done using the chemistry set within the APC.

**Comment by cinead** [comment]
I observed this error while attempting to batch craft a piece of short cordage in version 8aaf68f7.

```
Attempted a recipe with no available components!

comp_selection<item_comp> player::select_item_component(const std::vector<item_comp>&, int, inventory&, bool, const std::function<bool(const item&)>&, bool)

src/crafting.cpp
LINE 1539
```

CRAFTING.CPP:
```
        // Unlike with tools, it's a bad thing if there aren't any components available
        if( cmenu.entries.empty() ) {
            if( player_inv ) {
                if( has_trait( trait_DEBUG_HS ) ) {
                    selected.use_from = usage_from::player;
                    return selected;
                }
            }
            debugmsg( "Attempted a recipe with no available components!" );
            selected.use_from = usage_from::cancel;
            return selected;
        }

        cmenu.allow_cancel = can_cancel;
```

Repro in attached savefile by attempting to batch craft short cordage.

![short-cordage-piece-recipe-available](https://user-images.githubusercontent.com/16060264/86519138-c3046c80-bdec-11ea-894f-d189871a5e27.png)
![batch craft menu seems ok](https://user-images.githubusercontent.com/16060264/86519140-cd266b00-bdec-11ea-8ddd-7db26e9af7a4.png)

[Roncevaux-batchcraft-issue-37239.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4873672/Roncevaux-batchcraft-issue-37239.zip)

[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/4873695/debug.log)


Environment:
Build 8aaf68f7 on Linux64 Mint 19 Tara
./cataclysm from Gnome Terminal w terminal graphics

Mods:
> Dark Days Ahead
> Disable NPC Needs
> Magiclysm
> Rural-Only Mapgen
> Alternative Map Key
> Mythical Martial Arts
> No Rail Stations
> Graphical Overmap
> Graphical Overmap Magiclysm
> sees-player-icon, HitButton_iso

**Comment by RenechCDDA** [comment]
Looks to be resolved as of 05fe4bd44cdc39347ff09582ebf91c32db704ba2, tried with both saves. No debugmsg, no crashes.
